### PR TITLE
Fully implement lazy loading

### PR DESCRIPTION
**Pull request by tunglinn**
Parking lot data is now split by city, to allow for lazy loading.
Parking lots will only be loaded if the city is within view.
Added testing.

### CODE REVIEW

File: src/js/setUpSite.js [hunk 1]
```diff
@@ -159,18 +161,44 @@ const generateScorecard = (scoreCardEntry) => {
...
+ *
+ * @param cityId: E.g. `columbus-oh`.
+ * @param parkingLayer: GeoJSON layer with parking lot data
+ */
+const loadParkingLot = async (cityId, parkingLayer) => {
+  let load = true;
+  parkingLayer.getLayers().forEach((loadedCity) => {
+    if (loadedCity.feature.properties.id === cityId) {
+      load = false;
+    }
```
Comment (Eric-Arellano) [inline src/js/setUpSite.js:174]: I don't follow this. Instead, should the logic be 

```javascript
const city = parkingLayer.getLayers().find((city) => city.feature.properties.id == cityId);
load = city.feature.properties.<something>
```

Right now, this seems like it will always reload the same city no matter what, even if it was loaded before.

File: src/js/setUpSite.js [hunk 2]
```diff
@@ -159,18 +161,44 @@ const generateScorecard = (scoreCardEntry) => {
...
+ * Centers view to city.
  *
  * @param map: The Leaflet map instance.
+ * @param cityProperties: An object with a `layout` key (Leaflet value) and keys
+ *    representing the score card properties stored in `score-cards.json`.
+ */
+const snapToCity = async (map, cityProperties) => {
+  const { layer } = cityProperties;
+  map.fitBounds(layer.getBounds());
+};
```
Comment (Eric-Arellano) [inline src/js/setUpSite.js:192]:
```suggestion
};

```

File: src/js/setUpSite.js [hunk 3]
```diff
@@ -181,8 +209,16 @@ const setMapToCity = (map, cityId, cityProperties, toFitBounds = true) => {
...
-const setUpAutoScorecard = (map, cities) => {
-  map.on("moveend", () => {
+/**
+ * Pulls up scorecard for the city closest to the center of view.
+ * Also, loads parking lot data of any city in view.
+ *
+ * @param map: The Leaflet map instance.
+ * @param cities: Dictionary of cities with layer and scorecard info.
+ * @param parkingLayer: GeoJSON layer with parking lot data
+ */
```
Comment (Eric-Arellano) [inline src/js/setUpSite.js:219]: It'd be super helpful to split out refactors like this into a dedicated PR since this is such a large PR. Try to put any change that is safe to land on `main` even without lazy loading into that PR, so that this PR's diff is as minimal as possible.

File: tests/app/setUpSite.test.js [hunk 4]
```diff
@@ -246,3 +246,33 @@ test("scorecard pulls up city closest to center", async ({ page }) => {
   await expect(scoreCardTitle).toEqual("Birmingham, AL");
   await expect(cityToggleValue).toEqual("birmingham-al");
 });
+
+test("map only loads parking lots for visible cities", async ({ page }) => {
```
Comment (Eric-Arellano) [inline tests/app/setUpSite.test.js:250]: Great job testing this!

File: tests/scripts/base.test.js [hunk 5]
```diff
@@ -154,3 +158,68 @@ test.describe("updateCoordinates()", () => {
...
+      parkingLotData,
+      addDataPath
+    );
+    expect(result.error).toBeUndefined();
+    expect(result.value).toBeDefined();
+
+    await expectUpdatedFile(cityId, addDataPath);
+
+    await fs.rm(addDataPath);
+  });
```
Comment (Eric-Arellano) [inline tests/scripts/base.test.js:207]:
```suggestion
  });
  
```

File: scripts/base.js [hunk 6]
```diff
@@ -107,4 +107,53 @@ const updateCoordinates = async (
   return Ok("File updated successfully!");
 };
 
-export { Ok, Err, determineArgs, updateCoordinates };
+const updateParkingLots = async (
```
Comment (Eric-Arellano) [inline scripts/base.js:110]: This would be helpful to have some doctring describing what it does and the parameters

File: scripts/base.js [hunk 7]
```diff
@@ -107,4 +107,53 @@ const updateCoordinates = async (
...
+  if (!Array.isArray(newData.features) || newData.features.length !== 1) {
+    return Err(
+      "The script expects exactly one entry in `features` because you can only update one city at a time."
+    );
+  }
+
+  const newCoordinates = newData.features[0].geometry.coordinates;
+  const newGeometryType = newData.features[0].geometry.type;
+
+  if (addCity) {
```
Comment (Eric-Arellano) [inline scripts/base.js:135]: Better is to invert the conditional and early return, like:

```
if (!addCity) {
  return
}

const newFile = ...
```

See https://medium.com/swlh/return-early-pattern-3d18a41bba8 about this tip.

**Review by Eric-Arellano — COMMENTED** [review]
This looks great! Thank you for the well written code, including the error messages in `base.js`.

File: src/js/setUpSite.js (hunk 1 above)
Comment (tunglinn) [reply src/js/setUpSite.js:174]: `parkingLayer` is the layers of already loaded lots. 
So the logic right now is, if the city we want to load is already loaded (aka in `parkingLayer`), then don't load it.

File: src/js/setUpSite.js [hunk 8]
```diff
@@ -252,14 +278,19 @@ const setUpCitiesLayer = async (map) => {
...
-const setUpParkingLotsLayer = async (map) => {
-  const parkingLotsData = await import("../../data/parking-lots.geojson");
-  const lotsLayer = geoJSON(parkingLotsData, {
+/**
+ * Creates a GeoJSON layer to hold all parking lot polygons.
+ * Every cites' parking lots will be lazily added to this layer.
+ *
+ * @param map: The Leaflet map instance.
+ */
+const setUpParkingparkingLayer = async (map) => {
```
Comment (Eric-Arellano) [inline src/js/setUpSite.js:292]: Typo in the function name?

**Review by Eric-Arellano — APPROVED** [review]
🙌
